// @yunzhi/ng-router-testing/esm2015/lib/activated-route.stub.js
import { Subject } from 'rxjs';
import { tap } from 'rxjs/operators';
import { ActivatedRouteSnapshot, ActivationEnd } from '@angular/router';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
/**
 * 路由测试桩
 */
export class ActivatedRouteStub {
    /**
     * 当有新路由参数时，弹出ActivationEnd事件，供在服务中通过Router获取全局路由信息
     * 比如：实现回退功能时，需要对每个路由信息进行缓存
     * @param router 路由
     */
    constructor(router) {
        this.paramsSubject = new Subject();
        this.paramMapSubject = new Subject();
        this.queryParamsSubject = new Subject();
        this.queryParamMapSubject = new Subject();
        this.snapshot = {
            paramMap: {
                get: () => {
                    return 0;
                }
            }
        };
        this.router = router;
        this.params = this.paramsSubject.asObservable();
        this.paramMap = this.paramMapSubject.asObservable();
        this.queryParams = this.queryParamsSubject.asObservable();
        this.paramsSubject.pipe(tap(params => {
            const activatedRouteSnapshot = new ActivatedRouteSnapshot();
            activatedRouteSnapshot.params = params;
            this.router.emitEvent(new ActivationEnd(activatedRouteSnapshot));
        })).subscribe();
        // 注册导航后的回调
        this.router.registerNavigateCallbackFn((queryParams) => {
            this.queryParamsSubject.next(queryParams);
        });
    }
}
ActivatedRouteStub.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: ActivatedRouteStub, deps: [{ token: i1.Router }], target: i0.ɵɵFactoryTarget.Injectable });
ActivatedRouteStub.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: ActivatedRouteStub });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: ActivatedRouteStub, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.Router }]; } });
export class ParamMapImpl {
    constructor(values) {
        this.keys = [];
        this.values = {};
        for (const key in values) {
            if (values[key] !== undefined) {
                const value = values[key];
                this.keys.push(key);
                this.values[key] = value;
            }
        }
    }
    get(name) {
        return this.values[name];
    }
    getAll(name) {
        return this.values[name];
    }
    has(name) {
        return this.keys.indexOf(name) >= 0;
    }
}
