// @yunzhi/ng-router-testing/esm2015/lib/router-testing.module.js
import { NgModule } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { RouterTestingModule as AngularRouterTestingModule } from '@angular/router/testing';
import { ActivatedRouteStub } from './activated-route.stub';
import { RouterStub } from './router.stub';
import * as i0 from "@angular/core";
export class RouterTestingModule {
}
RouterTestingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterTestingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
RouterTestingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterTestingModule, imports: [AngularRouterTestingModule], exports: [AngularRouterTestingModule] });
RouterTestingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterTestingModule, providers: [
        { provide: ActivatedRoute, useClass: ActivatedRouteStub },
        { provide: Router, useClass: RouterStub },
    ], imports: [[AngularRouterTestingModule], AngularRouterTestingModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterTestingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [AngularRouterTestingModule],
                    providers: [
                        { provide: ActivatedRoute, useClass: ActivatedRouteStub },
                        { provide: Router, useClass: RouterStub },
                    ],
                    exports: [AngularRouterTestingModule]
                }]
        }] });
