// @yunzhi/ng-router-testing/esm2015/lib/router.stub.js
import { Subject } from 'rxjs';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class RouterStub {
    constructor() {
        /**
         * 事件源
         */
        this.eventsSubject = new Subject();
        this.navigateCallbackFns = [];
        this.events = this.eventsSubject.asObservable();
    }
    /**
     * 增加该函数的目的仅为满足Ionic对route-link指令的处理
     * @param commands 未知
     * @param navigationExtras 未知
     */
    createUrlTree(commands, navigationExtras = {}) {
        return {
            root: null,
            queryParams: null,
            fragment: null,
            queryParamMap: null,
            toString() {
                return '';
            }
        };
    }
    /**
     * 弹出新事件
     * @param event 事件
     */
    emitEvent(event) {
        this.eventsSubject.next(event);
    }
    navigateByUrl(url, extras) {
        console.log('接收到了跳转请求', url, extras);
        return Promise.resolve(true);
    }
    /**
     * 跳转
     * @param commands 参数
     * @param extras 扩展属性
     */
    navigate(commands, extras) {
        this.navigateCallbackFns.forEach(callback => callback(extras.queryParams));
        return Promise.resolve(true);
    }
    /**
     * 增加该函数的目的仅为满足Ionic对route-link指令的处理
     * @param url 未知
     */
    serializeUrl(url) {
        return '';
    }
    /**
     * 注册导航返回的回调
     * @param fn 回调函数
     */
    registerNavigateCallbackFn(fn) {
        this.navigateCallbackFns.push(fn);
    }
}
RouterStub.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterStub, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
RouterStub.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterStub });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.5", ngImport: i0, type: RouterStub, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return []; } });
